pkgs/development/interpreters/python/wrapper.nix

```diff
@@ -1,59 +1,76 @@
+/* Create a Python executable that knows about additional packages. */
+
 { stdenv, python, buildEnv, makeWrapper
 , extraLibs ? []
 , extraOutputsToInstall ? []
 , postBuild ? ""
 , ignoreCollisions ? false }:
 
-# Create a python executable that knows about additional packages.
+with stdenv.lib;
+
 let
-  recursivePthLoader = import ../../python-modules/recursive-pth-loader/default.nix { stdenv = stdenv; python = python; };
-  env = let
-    paths = stdenv.lib.closePropagation (extraLibs ++ [ python recursivePthLoader ] ) ;
-  in buildEnv {
-    name = "${python.name}-env";
+  recursivePthLoader = import ../../python-modules/recursive-pth-loader/default.nix {
+    inherit stdenv python;
+  };
+  inputs = extraLibs ++ [ python recursivePthLoader ];
+  # https://git.io/vFV56
+  pickOutputs = drv:
+    if (drv.outputUnspecified or false) && (drv.meta.outputsToInstall or null) != null
+    then (map (out: getOutput out drv) drv.meta.outputsToInstall)
+    else [ drv ]
+    ++ filter isNull (map (out: getOutput out drv) extraOutputsToInstall);
+  closePropagation' = drvs:
+    map (drv: drv // { outputUnspecified = true; }) (closePropagation drvs);
+  paths = concatLists (map pickOutputs (closePropagation' inputs));
+  self = buildEnv {
+    inherit ignoreCollisions extraOutputsToInstall paths;
+    inherit (python) meta;
 
-    inherit paths;
-    inherit ignoreCollisions extraOutputsToInstall;
+    name = "${python.name}-env";
+    buildInputs = [ makeWrapper ];
 
     postBuild = ''
-      . "${makeWrapper}/nix-support/setup-hook"
-
-      if [ -L "$out/bin" ]; then
-          unlink "$out/bin"
+      if [ -L $out/bin ]; then
+        unlink $out/bin
       fi
-      mkdir -p "$out/bin"
-
-      for path in ${stdenv.lib.concatStringsSep " " paths}; do
-        if [ -d "$path/bin" ]; then
-          cd "$path/bin"
-          for prg in *; do
-            if [ -f "$prg" ]; then
-              rm -f "$out/bin/$prg"
```

Comment:
This is the only functional change between the previous `python3.buildEnv` and this one: if `$path/bin/$prg` is a non-executable file (which is arguably a packaging mistake), `$out/bin/$prg` is not deleted anymore.